### PR TITLE
Bilateral trade realization

### PR DESCRIPTION
**Pull request by caviddhen**
## :bird: Purpose of this PR :bird:

- Briefly explain the purpose of this pull request

## :wrench: Checklist for PR creator :wrench:

- [x] Labeling pull request correctly [from the label list](https://github.com/magpiemodel/magpie/labels).

* Low risk : Simple bugfixes (missing files, updated documentation, typos) or Start/output scripts
* Medium risk : New realization / Changes to existing realization / Other changes which don't modify default.cfg
* High risk : New input files (if cfg$input is changed in default.cfg) / Modification to core model (eg. changes in equations, calculations, introduction of new sets etc.) / Other changes in default.cfg

- [x] Providing additional information based on PR label

* Low risk : No new model run needed.
* Medium risk : Default run based on the current version of the fork from which PR is made
* High risk
  * Default run from the current develop branch
  * Default run based on the current version of the fork from which PR is made

- [x] :chart_with_downwards_trend: Performance loss/gain from current default behavior :chart_with_upwards_trend:
  * Current develop branch default : 0.45 hours
  * This PR's default :  0.44 hours

- [x] Added changes to `CHANGELOG.md`
- [x] Compilation check (model starts without compilation errors - use `gams main.gms action=c` in model folder for testing).
- [x] No hard coded numbers and cluster/country/region names.
- [x] The new code doesn't contain declared but unused parameters or variables.
- [x] Where relevant, In-code comments added including documentation comments.
- [x] Made sure that documentation created with [`goxygen`](https://github.com/pik-piam/goxygen) is okay (use `goxygen::goxygen()` for testing).
- [x] Changes to [`magpie4`](https://github.com/pik-piam/magpie4) R library for post processing of model output (ideally backward compatible).
- [x] Self-review of my own code.
- [x] In case of updated cellular input tgz file in default.cfg: scenario_config.csv has been updated accordingly (rcp1p9, rcp2p6 etc)
- [x] For high risk runs: validation of major model indicators - Land-use, emissions, food prices, Tau. Not high risk, but interesting to see. 

Resources_Land_Cover_Cropland (13)

Trade_Net_Trade (5)

Trade_Net_Trade_Crops (1)

Trade_Net_Trade_Livestock_products (2)

Emissions_CO2_Land_Land_use_Change (10)

Productivity_Landuse_Intensity_Indicator_Tau (10)

## :warning: Additional notes or warnings :warning:
NA

## :rotating_light: Checklist for RSE reviewer :rotating_light:

- [ ] PR is labeled correctly.
- [ ] `CHANGELOG` is updated correctly
- [ ] No hard coded numbers and cluster/country/region names.
- [ ] No unnecessary increase in module interfaces
- [ ] All required updates in interfaces (if any) have been properly adressed in the module contracts
- [ ] In-code comments and documentation comments are satisfactory.
- [ ] model behavior/performance is satisfactory.
- [ ] Requested changes (if any) were applied correctly

## :rotating_light: Checklist for MAgPIE reviewer :rotating_light:

- [ ] PR is labeled correctly.
- [ ] `CHANGELOG` is updated correctly
- [ ] No hard coded numbers and cluster/country/region names.
- [ ] Changes to the model are scientifically sound
- [ ] In-code comments and documentation comments are satisfactory.
- [ ] model behavior/performance is satisfactory.
- [ ] Requested changes (if any) were applied correctly

### CODE REVIEW

**Comment by bodirsky** [comment]
you have a packaging and local trade plot included above, i guess this should not be part of this PR?

File: modules/21_trade/selfsuff_reduced_bilateral22/realization.gms [hunk 1]
```diff
@@ -0,0 +1,34 @@
+*** |  (C) 2008-2021 Potsdam Institute for Climate Impact Research (PIK)
```
Comment (bodirsky) [inline modules/21_trade/selfsuff_reduced_bilateral22/realization.gms:1]: documentation doesnt mention bilateral trade except for in the limitations

File: modules/42_water_demand/agr_sector_aug13/input.gms [hunk 2]
```diff
@@ -33,7 +33,7 @@ s42_env_flow_base_fraction         Fraction of available water that is reserved
 * 									(determined in the file EFR_protection_policy.csv)
 s42_env_flow_fraction              Fraction of available water that is reserved for under protection policies (1) / 0.2 /
 s42_pumping                        Switch to activate pumping cost settings (1) / 0 /
-s42_multiplier_startyear           Year from which pumping costs multiplier will be implemented (1) / 1995 /
+s42_multiplier_startyear           Year from which pumping costs multiplier will be implemented (1) / 0 /
```
Comment (bodirsky) [inline modules/42_water_demand/agr_sector_aug13/input.gms:36]: this happened in my case, too...
where does this come from?
guess should be reverwed..

File: modules/42_water_demand/all_sectors_aug13/input.gms [hunk 3]
```diff
@@ -37,7 +37,7 @@ s42_env_flow_base_fraction         Fraction of available water that is reserved
 *                                                                   EFR_protection_policy.csv)
 s42_env_flow_fraction              Fraction of available water that is reserved for under protection policies (1) / 0.2 /
 s42_pumping                        Switch to activate pumping cost settings (1) / 0 /
-s42_multiplier_startyear           Year from which pumping costs multiplier will be implemented (1) / 1995 /
+s42_multiplier_startyear           Year from which pumping costs multiplier will be implemented (1) / 0 /
```
Comment (bodirsky) [inline modules/42_water_demand/all_sectors_aug13/input.gms:40]: also here, why is this changed? seems to be some automatic script that does that

**Review by bodirsky — CHANGES_REQUESTED** [review]

File: modules/21_trade/selfsuff_reduced_bilateral22/equations.gms [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ v21_cost_tariff_reg(h2,k_trade) =g=
+  sum(h_im, i21_trade_tariff(h2,h_im,k_trade))
+ *sum(supreg(h2,i2), vm_prod_reg(i2,k_trade)-vm_supply(i2,k_trade));
+
+* Trade margins (tariffs) costs assigned currently to exporting region. Margins at region level 
+q21_costs_margins(i2,k_trade)..
+ v21_cost_transport_reg(i2,k_trade) =g=
+  sum(i_im, i21_trade_margin(i2,i_im,k_trade))* sum(i_im, v21_trade(i2,i_im,k_trade));
+
+* Regional trade costs are the costs for each region aggregated over all the tradable commodities.
```
[the remark this answers is not in the document]
Comment (caviddhen) [reply modules/21_trade/selfsuff_reduced_bilateral22/equations.gms:85]: do you mean to have costs and tariffs at iso level? so not aggregated?

File: modules/21_trade/input/files [hunk 5]
```diff
@@ -5,6 +5,10 @@ f21_trade_balanceflow.cs3
...
 f21_trade_tariff.cs3
 f21_trade_tariff_export.cs3
-f21_trade_tariff_import.cs3       
+f21_trade_tariff_import.cs3     
+f21_trade_margin_bilat.cs5
+f21_trade_tariff_bilat.cs5
+f21_trade_tariff_export_bilat.cs5
+f21_trade_tariff_import_bilat.cs5  
 f21_trade_bal_reduction.cs3
 f21_trade_balance.cs3
```
[the remark this answers is not in the document]
Comment (caviddhen) [reply modules/21_trade/input/files:9]: done

File: modules/21_trade/selfsuff_reduced_bilateral22/declarations.gms [hunk 6]
```diff
@@ -0,0 +1,57 @@
+*** |  (C) 2008-2021 Potsdam Institute for Climate Impact Research (PIK)
```
[the remark this answers is not in the document]
Comment (caviddhen) [reply modules/21_trade/selfsuff_reduced_bilateral22/declarations.gms:1]: I did this but it seems like git is already tracking anything named the same. 

The main additions are q21_trade_bilat, q21_costs_tariffs, q21_costs_margins

File: modules/21_trade/selfsuff_reduced_bilateral22/declarations.gms [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+*** |  authors, and contributors see CITATION.cff file. This file is part
+*** |  of MAgPIE and licensed under AGPL-3.0-or-later. Under Section 7 of
+*** |  AGPL-3.0, you are granted additional permissions described in the
+*** |  MAgPIE License Exception, version 1.0 (see LICENSE file).
+*** |  Contact: magpie@pik-potsdam.de
+
+parameters
+ i21_trade_bal_reduction(t_all,k_trade)         Trade balance reduction (1)
+ i21_trade_margin(i_ex,i_im,k_trade)            Trade margins (USD05MER per tDM)
```
[the remark this answers is not in the document]
Comment (caviddhen) [reply modules/21_trade/selfsuff_reduced_bilateral22/declarations.gms]: done

File: modules/21_trade/selfsuff_reduced_bilateral22/equations.gms [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+q21_trade_bilat(h2,k_trade)..
+ sum(supreg(h2, i2), vm_prod_reg(i2, k_trade)) =g= sum(supreg(h2,i2), (vm_supply(i2, k_trade) -
+                              sum(i_ex, v21_trade(i_ex, i2, k_trade))  + sum(i_im, v21_trade(i2, i_im, k_trade))));
+
+*' For non-tradable commodites, the regional supply should be larger or equal to the regional demand.
+ q21_notrade(h2,k_notrade)..
+  sum(supreg(h2,i2),vm_prod_reg(i2,k_notrade)) =g= sum(supreg(h2,i2), vm_supply(i2,k_notrade));
+
+
+*' The following equation indicates the regional trade constraint for the self-sufficiency pool.
```
[the remark this answers is not in the document]
Comment (caviddhen) [reply modules/21_trade/selfsuff_reduced_bilateral22/equations.gms:25]: yes correct

File: modules/21_trade/selfsuff_reduced_bilateral22/equations.gms [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+*** |  of MAgPIE and licensed under AGPL-3.0-or-later. Under Section 7 of
+*** |  AGPL-3.0, you are granted additional permissions described in the
+*** |  MAgPIE License Exception, version 1.0 (see LICENSE file).
+*** |  Contact: magpie@pik-potsdam.de
+
+*' @equations
+*' In the comparative advantage pool, the only active constraint is that the global supply is larger or equal to demand.
+*' This means that production can be freely allocated globally based on comparative advantages.
+
+ q21_trade_glo(k_trade)..
```
[the remark this answers is not in the document]
Comment (caviddhen) [reply modules/21_trade/selfsuff_reduced_bilateral22/equations.gms:12]: still required for the balance flow

**Review by bodirsky — APPROVED** [review]

**Review by bodirsky — COMMENTED** [review]
At the moment, the s21_trade_tariff is a pure sensitivity analysis.

It would be nice to also have the option of fading out tariffs, as a trade-lib scenario.

**Review by bodirsky — CHANGES_REQUESTED** [review]
the timing of the trade-tariff switch (2010) is against our convention that makes scenario switches only active after 2020

**Comment by caviddhen** [comment]
> the timing of the trade-tariff switch (2010) is against our convention that makes scenario switches only active after 2020

where did you see this? trade tariffs used to not have a time dimension, so it was either completely off or on for all timesteps

**Review by bodirsky — APPROVED** [review]

File: main.gms [hunk 10]
```diff
@@ -244,12 +249,12 @@ $setglobal production  flexreg_apr16
 $setglobal residues  flexreg_apr16
 $setglobal processing  substitution_may21
 
-$setglobal trade  selfsuff_reduced
+$setglobal trade  selfsuff_reduced_bilateral22
```
Comment (tscheypidi) [inline main.gms:252]: please do not commit changes in main.gms

File: modules/21_trade/selfsuff_reduced_bilateral22/declarations.gms [hunk 11]
```diff
@@ -0,0 +1,62 @@
+*** |  (C) 2008-2021 Potsdam Institute for Climate Impact Research (PIK)
```
Comment (tscheypidi) [inline modules/21_trade/selfsuff_reduced_bilateral22/declarations.gms:1]: Please check that copyright is updated in all files to 2008-2023

```suggestion
*** |  (C) 2008-2023 Potsdam Institute for Climate Impact Research (PIK)
```

File: modules/21_trade/selfsuff_reduced_bilateral22/equations.gms [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+ + sum(ct,f21_trade_balanceflow(ct,k_trade));
+
+*' Distributing the global excess demand to exporting regions is based on regional export shares [@schmitz_trading_2012].
+*' Export shares are derived from FAO data (see @schmitz_trading_2012 for details). They are 0 for importing regions.
+
+ q21_excess_supply(h2,k_trade)..
+ v21_excess_prod(h2,k_trade) =e=
+ v21_excess_dem(k_trade)*sum(ct,f21_exp_shr(ct,h2,k_trade));
+
+* Trade tariffs are associated with exporting regions. They are dependent on net exports and tariff levels.
```
Comment (tscheypidi) [inline modules/21_trade/selfsuff_reduced_bilateral22/equations.gms:68]: Shouldn't this be part of the goxygen documentation as well?
```suggestion
*' Trade tariffs are associated with exporting regions. They are dependent on net exports and tariff levels.
```
Same question/suggestion for the following comments.

**Review by tscheypidi — CHANGES_REQUESTED** [review]
Looks mostly ready. Just some things to be considered/changed:

* Please check all copyright statements and make sure that they range till 2023
* please check the goxygen documentation. Code looks a bit like some simple comments should actually be goxygen comments. Apart from that it would be important to check whether the documentation for the realization looks good.
* 169kB is quite a lot for such a simple figure. As it is now already part of the git history a change would only further increase the repository size but please try to avoid such file sizes in the future if possible

**Comment by caviddhen** [comment]
> Looks mostly ready. Just some things to be considered/changed:
> 
> * Please check all copyright statements and make sure that they range till 2023
> * please check the goxygen documentation. Code looks a bit like some simple comments should actually be goxygen comments. Apart from that it would be important to check whether the documentation for the realization looks good.
> * 169kB is quite a lot for such a simple figure. As it is now already part of the git history a change would only further increase the repository size but please try to avoid such file sizes in the future if possible

Thanks Jan! adjusted now! and the documentation looks ok too. (and sorry, the image had copied over from the last realization, as this one builds on it, noted for the future).

**Review by tscheypidi — APPROVED** [review]

**Pull request merged**